### PR TITLE
Avoid shell-ing out to commands in etcd migrate utility

### PR DESCRIPTION
**Pull request by dims**
It's ok to shell out to etcdctl etc, but not things like "cp" and "mv"
as this stops us from moving to distroless in the future

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by dims** [comment]
/hold

for reviews

**Comment by jpbetz** [comment]
Was lazy of me not to fix this when migrating from bash to go. Thanks for fixing @dims!

/lgtm

**Comment by dims** [comment]
/hold

for test failures and windows typecheck failures.

File: cluster/images/etcd/migrate/migrator.go [hunk 1]
```diff
@@ -159,7 +158,7 @@ func (m *Migrator) rollbackEtcd3MinorVersion(current *EtcdVersionPair, target *E
 	if err != nil {
 		return nil, err
 	}
-	err = exec.Command("mv", m.dataDirectory.path, backupDir).Run()
+	err = os.Rename(m.dataDirectory.path, backupDir)
```
Comment (liggitt) [inline cluster/images/etcd/migrate/migrator.go:161]: mv handles different disks... does os.Rename?

File: cluster/images/etcd/migrate/util_others.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+*/
+
+package main
+
+import (
+	"github.com/mrunalp/fileutils"
+)
+
+func copyDirectory(source string, dest string) error {
+	return fileutils.CopyDirectory(source, dest)
```
Comment (liggitt) [inline cluster/images/etcd/migrate/util_others.go:26]: I'm not thrilled with the level of test coverage for this (none in https://github.com/mrunalp/fileutils)

Our unit test is pretty thin (just makes sure a file exists). @jpbetz, are there weird things in the data dir we need to handle specially (links, sockets, files with weird permissions, etc)?

Comment (jpbetz) [reply cluster/images/etcd/migrate/util_others.go:26]: No links, sockets or strange permissions, fortunately. The data directory contains `.wal` files, `.snap` files, the main `db` file and some temporary `db.tmp` and `.snap.db` files for specific operations. They are all normal files.

Comment (liggitt) [reply cluster/images/etcd/migrate/util_others.go:26]: ok. does this produce files with the same permissions as `cp`? I don't know how much we rely on file permissions here to protect the etcd data

File: cluster/images/etcd/migrate/migrator.go (hunk 1 above)
Comment (jpbetz) [reply cluster/images/etcd/migrate/migrator.go:161]: Looks like the answer is no, although I haven't tried it. There is a open issues about this: https://github.com/cloudfoundry/bosh-utils/issues/10

Comment (dims) [reply cluster/images/etcd/migrate/migrator.go:161]: @liggitt this is running inside a container right? do we need to worry about multiple disks?

Comment (liggitt) [reply cluster/images/etcd/migrate/migrator.go:161]: I'm really unfamiliar with how it is used... is all the data internal only or are primary/backup volumes mounted into the container?

File: cluster/images/etcd/migrate/util_others.go (hunk 2 above)
Comment (dims) [reply cluster/images/etcd/migrate/util_others.go:26]: some background, there's a reference in vendor/github.com/opencontainers/runc/libcontainer/rootfs_linux.go already so i picked this impl.

File: cluster/images/etcd/migrate/migrator.go (hunk 1 above)
Comment (dims) [reply cluster/images/etcd/migrate/migrator.go:161]: @liggitt Data never leaves the container, there are no volumes mounted, the mainfest call is here:
https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/manifests/etcd.manifest#L27

the `migrate-if-needed.sh` is here:
https://github.com/kubernetes/kubernetes/blob/master/cluster/images/etcd/migrate-if-needed.sh

The `DATA_DIRECTORY` stays within the container:
https://cs.k8s.io/?q=DATA_DIRECTORY&i=nope&files=&repos=kubernetes/kubernetes

The backup directory is the same as the original directory path with a suffix:
https://github.com/kubernetes/kubernetes/blob/master/cluster/images/etcd/migrate/data_dir.go#L82

Comment (liggitt) [reply cluster/images/etcd/migrate/migrator.go:161]: Ok, thanks

**Comment by dims** [comment]
thanks @liggitt

**Comment by dims** [comment]
to try this out we need to publish an image and edit etcd.manifest

```
diff --git cluster/gce/manifests/etcd.manifest cluster/gce/manifests/etcd.manifest
index 99f08d1ff195..06b2f8224ff0 100644
--- cluster/gce/manifests/etcd.manifest
+++ cluster/gce/manifests/etcd.manifest
@@ -15,7 +15,7 @@
 "containers":[
     {
     "name": "etcd-container",
-    "image": "gcr.io/kubernetes-development-244305/etcd-amd64:3.4.7-2",
+    "image": "{{ pillar.get('etcd_docker_repository', 'k8s.gcr.io/etcd') }}:{{ pillar.get('etcd_docker_tag', '3.4.7-0') }}",
     "resources": {
       "requests": {
         "cpu": {{ cpulimit }}
@@ -24,7 +24,7 @@
     "command": [
               "/bin/sh",
               "-c",
-              "set -o errexit; if [ -e /usr/local/bin/migrate-if-needed.sh ]; then /usr/local/bin/migrate-if-needed.sh 1>>/var/log/etcd{{ suffix }}.log 2>&1; fi; exec /usr/local/bin/etcd --name etcd-{{ hostname }} --listen-peer-urls {{ etcd_protocol }}://{{ host_ip }}:{{ server_port }} --initial-advertise-peer-urls {{ etcd_protocol }}://{{ hostname }}:{{ server_port }} --advertise-client-urls {{ etcd_apiserver_protocol }}://127.0.0.1:{{ port }} --listen-client-urls {{ etcd_apiserver_protocol }}://{{ listen_client_ip }}:{{ port }} {{ quota_bytes }} --data-dir /var/etcd/data{{ suffix }} --initial-cluster-state {{ cluster_state }} --initial-cluster {{ etcd_cluster }} {{ etcd_creds }} {{ etcd_apiserver_creds }} {{ etcd_extra_args }} 1>>/var/log/etcd{{ suffix }}.log 2>&1"
+              "if [ -e /usr/local/bin/migrate-if-needed.sh ]; then /usr/local/bin/migrate-if-needed.sh 1>>/var/log/etcd{{ suffix }}.log 2>&1; fi; exec /usr/local/bin/etcd --name etcd-{{ hostname }} --listen-peer-urls {{ etcd_protocol }}://{{ host_ip }}:{{ server_port }} --initial-advertise-peer-urls {{ etcd_protocol }}://{{ hostname }}:{{ server_port }} --advertise-client-urls {{ etcd_apiserver_protocol }}://127.0.0.1:{{ port }} --listen-client-urls {{ etcd_apiserver_protocol }}://{{ listen_client_ip }}:{{ port }} {{ quota_bytes }} --data-dir /var/etcd/data{{ suffix }} --initial-cluster-state {{ cluster_state }} --initial-cluster {{ etcd_cluster }} {{ etcd_creds }} {{ etcd_apiserver_creds }} {{ etcd_extra_args }} 1>>/var/log/etcd{{ suffix }}.log 2>&1"
             ],
     "env": [
       { "name": "TARGET_STORAGE",

```

**Comment by dims** [comment]
switching to https://github.com/kubernetes/kubernetes/pull/91171

**Pull request closed without merging**